operators/nftopology/controllers/nftopology_controller.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	pvv1alpha1 "github.com/GoogleContainerTools/kpt/porch/controllers/packagevariants/api/v1alpha1"
+	pvsv1alpha2 "github.com/GoogleContainerTools/kpt/porch/controllers/packagevariantsets/api/v1alpha2"
+	"github.com/go-logr/logr"
+	deployv1alpha1 "github.com/nephio-project/api/nf_deployments/v1alpha1"
+	reqv1alpha1 "github.com/nephio-project/api/nf_requirements/v1alpha1"
+)
+
+const (
+	FREE5GC            string = "free5gc"
+	WORKLOADAPIVERSION string = "infra.nephio.org/v1alpha1"
+	WORKLOADKIND       string = "WorkloadCluster"
+)
+
+// NFTopologyReconciler reconciles a NFTopology object
+type NFTopologyReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	l      logr.Logger
+}
+
+func AllUPFsDeployed(neighborMap map[string][]reqv1alpha1.NFInstance, nfInst *reqv1alpha1.NFInstance, nfDeployed *deployv1alpha1.NFDeployed, deployedInstMap map[string]int) bool {
+
+	expected := 0
+	count := 0
+	if neighborList, ok := neighborMap[nfInst.Name]; !ok {
+		return false
+	} else {
+		// create a map for fast lookup
+		neighborLookupMap := map[string]struct{}{}
+		for _, neighbor := range neighborList {
+			if neighbor.NFTemplate.NFType == reqv1alpha1.NFTypeUPF {
+				neighborLookupMap[neighbor.Name] = struct{}{}
+				expected++
+			}
+		}
+
+		for neighborInstName, idx := range deployedInstMap {
+			if _, ok := neighborLookupMap[GetNFInstanceNameFromNFDeployedName(neighborInstName)]; ok {
+				if nfDeployed.Spec.NFInstances[idx].NFType == string(reqv1alpha1.NFTypeUPF) {
+					count++
+				}
+			}
+		}
+	}
+
+	if expected == count {
+		return true
+	} else {
+		return false
+	}
+}
+
+/*
+ * BuildPVS: builds a PVS object from NFTopology's NFInstance
+ * NFClass needed for package reference
+ */
+func BuildPVS(nfInst *reqv1alpha1.NFInstance, nfClassObj *reqv1alpha1.NFClass) *pvsv1alpha2.PackageVariantSet {
```

Comment:
Is there a reason these (AllUPFsDeployed, BuildPVS, etc.) are exported (capitalized)? Are they called by another module? If not, they should be local (lowercase first letter).